src/main/java/org/truffleruby/language/SetTopLevelBindingNode.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2018 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.language;
+
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.object.DynamicObject;
+import org.truffleruby.Layouts;
+import org.truffleruby.RubyContext;
+import org.truffleruby.core.module.ModuleFields;
+
+public class SetTopLevelBindingNode extends RubyNode {
+
+    @Child private RubyNode child;
+
+    public SetTopLevelBindingNode(RubyNode child) {
+        this.child = child;
+    }
+
+    @Override
+    public Object execute(VirtualFrame frame) {
+        final RubyContext context = getContext();
+        final ModuleFields fields = Layouts.MODULE.getFields(context.getCoreLibrary().getObjectClass());
+
+        final RubyConstant previousConstant = fields.getConstant("TOPLEVEL_BINDING");
+        final Object previousValue = previousConstant == null ? null : previousConstant.getValue();
+
+        final DynamicObject binding = Layouts.BINDING.createBinding(context.getCoreLibrary().getBindingFactory(), frame.materialize());
+        fields.setConstant(context, this, "TOPLEVEL_BINDING", binding);
+
+        try {
+            return child.execute(frame);
+        } finally {
+            if (previousValue == null) {
+                fields.removeConstant(context, this, "TOPLEVEL_BINDING");
+            } else {
+                fields.setConstant(context, this, "TOPLEVEL_BINDING", previousValue);
+            }
+        }
```

Comment:
So this re-set the TOPLEVEL_BINDING for every file loaded?
That's bad, as loading any file will deopt many constant lookups.
It also doesn't seem thread-safe.

Reply:
Isn't there a global require lock?

Reply:
And surely requiring any file is going to cause quite a lot of depots due to new classes and methods anyway?

Reply:
No, the require lock is per-file.
Yes, indeed, any file defining a new class/module or constant will invalidate some constant lookups (but could be under a namespace where the impact is much smaller than Object constants).